### DISCUSSION TITLE
.NET 6/API7

### DISCUSSION THREAD

**Discussion by goaaats** [goatcorp/Dalamud#781] [Announcements]
Hey everyone, we upgraded the Dalamud core library and the runtime we are shipping to .NET 6 for staging releases.

# Breaking changes
Please note that there are [breaking changes in .NET 6](https://docs.microsoft.com/en-us/dotnet/core/compatibility/6.0#core-net-libraries). Check your plugins - if you come across a regression that isn't listed, please let us know.

## Breaking changes not listed in MSFT docs
* The .NET 6 runtime **does not tolerate invalid calling conventions anymore.** You have to use the right calling convention(ALWAYS FastCall in Dalamud, never ThisCall or any others) when specifying delegates for native functions, or leave it up to the runtime.

# Testing
Open the dev menu with `/xldev`, click Dalamud, click "open branch switcher", choose "stg", then restart.

# Upgrade guide
Change your plugin target framework to `net6.0-windows` or `net6.0`. Make sure to update all Microsoft framework-versioned nuget packages. Work through the breaking changes.

Please report anything else you run into here.

# Publish with the new publishing system
Our new plugin publishing system that build your plugin for you is pretty much done, and we encourage you to switch to it as soon as possible, the .NET 6 being a good opportunity. With 6.3, the old system will be decommissioned. It's super easy to switch and will make your workflow a lot easier. If you need help, feel free to reach out. [Check here for a short guide](https://github.com/goatcorp/DalamudPluginsD17/blob/main/README.md).

# Timeline
* Now to 6.2: Testing period - we encourage you to submit versions of your plugins with API level 7 to `testing/net6` on the new distribution repo, these will be transferred over once the update hits. The .NET 6 testing branch will load API6 and API7 plugins for now in order to not break your workflow, as most .NET 5 plugins seem to be compatible.
* 08/18: End of breaking changes
* 6.2 release: official switch to .NET 6

**Comment by anna-is-cute** [comment]
Some questions regarding updating now that breaking changes are over:

1. Is <https://github.com/goatcorp/dalamud-distrib/blob/main/stg/latest.zip> the right set of assemblies for updating to net6 and API 7?
2. Will plugins submitted to `testing/net6` be loaded as `stable` once the API 7 update happens?
3. When will there be a DalamudPackager update to actually put API 7 in the manifest without manually specifying it?
4. Are images no longer required to be included in the `csproj`? Or to put it another way, do images only need to be included in `DalamudPluginsD17` now?
